File: java-ladidol/src/com/feng/newline/dp/stock/L188买卖股票的最佳时机IV.java
package com.feng.newline.dp.stock;

/**
 * @projectName: leetcode
 * @package: com.feng.newline.dp
 * @className: L188买卖股票的最佳时机IV
 * @author: Ladidol
 * @description: 给定一个整数数组prices ，它的第 i 个元素prices[i] 是一支给定的股票在第 i 天的价格。
 * 设计一个算法来计算你所能获取的最大利润。你最多可以完成 k 笔交易。
 * 注意：你不能同时参与多笔交易（你必须在再次购买前出售掉之前的股票）。
 * 示例 1：
 * <p>
 * 输入：k = 2, prices = [2,4,1]
 * 输出：2
 * 解释：在第 1 天 (股票价格 = 2) 的时候买入，在第 2 天 (股票价格 = 4) 的时候卖出，这笔交易所能获得利润 = 4-2 = 2 。
 * 示例 2：
 * <p>
 * 输入：k = 2, prices = [3,2,6,5,0,3]
 * 输出：7
 * 解释：在第 2 天 (股票价格 = 2) 的时候买入，在第 3 天 (股票价格 = 6) 的时候卖出, 这笔交易所能获得利润 = 6-2 = 4 。
 * 随后，在第 5 天 (股票价格 = 0) 的时候买入，在第 6 天 (股票价格 = 3) 的时候卖出, 这笔交易所能获得利润 = 3-0 = 3 。
 * @date: 2022/9/3 20:28
 * @version: 1.0
 */
public class L188买卖股票的最佳时机IV {
    // 二维dp，根据123买卖股票III推演过来，这题的每天就有2k+1个状态的，出去什么都不干就还剩2k个状态了。
    class Solution {
        public int maxProfit(int k, int[] prices) {
            if (prices.length == 0) return 0;
            // 奇数是买入，偶数是卖出。
            int[][] dp = new int[prices.length][2 * k + 1];

            //初始化
            for (int j = 1; j < 2 * k; j += 2) {
                dp[0][j] = -prices[0];
            }

            for (int i = 1; i < prices.length; i++) {
                for (int j = 1; j < 2 * k; j += 2) {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - 1] - prices[i]);//买入
                    dp[i][j + 1] = Math.max(dp[i - 1][j + 1], dp[i - 1][j] + prices[i]);//卖出
                }
            }
            return dp[dp.length - 1][2 * k];
        }
    }

    // 有些为了方便表示用了三维数组来做，其实是一样的。
}

File: golang-devildyw/hard/L188买卖股票的最佳时机IV.go
package main

func main() {

}

func maxProfitIV(k int, prices []int) int {
	// 有点像上一道题 不过是把2换成k了

	// 需要注意一个边界条件 那就是当k>=n/2的时候 因为买卖一次至少需要2天 如果超过就退化成了无限次买卖了 就跟买卖股票的最佳时机II 一样了
	n := len(prices)

	// if (k >= n/2) {
	//     //使用买卖股票最佳时机II的做法
	// }

	// dp[i][0] dp[i][1] 代表着第i笔买入或者交易的利润 因为会出现i-1的情况 所以这里增加一个位置来处理
	dp := make([][]int, k+1)
	for i := range k + 1 {
		dp[i] = make([]int, 2)
		// 初始化
		dp[i][0] = -prices[0]
	}
	dp[0][0] = 0

	// 状态转移方程
	for i := 1; i < n; i++ {
		// 随着第i天的变化 这k组不同状态的利润也会随之改变
		for j := 1; j <= k; j++ {
			// 第j笔买入时的利润 要么保持不变 要么就是上一笔交易的利润减去当前第i天的股票价格 取利润最大值
			dp[j][0] = max(dp[j][0], dp[j-1][1]-prices[i])
			// 第j笔卖出时的利润 要么保持不变 要么就是当前股票的价格+之前买入的利润 取最大值
			dp[j][1] = max(dp[j][1], prices[i]+dp[j][0])
		}
	}

	return dp[k][1]

}
